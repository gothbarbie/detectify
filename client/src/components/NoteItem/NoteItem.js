import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import moment from 'moment'
import { connect } from 'react-redux'

import { showModalAndSetDeleteNoteIdThunk } from '../../actions/modalActions'
import { updateNoteThunk } from '../../actions/notesActions'

import EditButton from '../EditButton'
import DeleteNoteButton from '../DeleteNoteButton'
import NoteHeader from './NoteHeader'
import EditText from '../EditText'

const Article = styled.article`
  border-top: 1px solid #999999;
`

const Main = styled.main`
  background-color: #f4f9ff;
  padding: 1rem;
`

const Timestamp = styled.div`
  font-size: 1.1rem;
  color: #222;
  font-weight: 400;
`

const Category = styled.div`
  font-size: 1.3rem;
  color: #6b85a4;
  font-weight: 700;
`

const Content = styled.textarea`
  padding: 1rem 0;
  font-size: 1.2rem;
  color: #444;
  border: none;
  width: 100%;
  background: transparent;
`

class NoteItem extends Component {
  state = {
    open: false,
    editCategory: false,
    editTitle: false,
    title: this.props.note.title,
    content: this.props.note.content,
    category: this.props.note.category,
  }

  toggle = () => {
    this.setState(prevState => ({
      open: !prevState.open,
    }))
  }

  setEditCategory = () =>
    this.setState({
      editCategory: true,
    })

  setEditTitle = () => this.setState({ editTitle: true })

  onChange = event => {
    this.setState({ [event.target.name]: event.target.value })
  }

  onBlur = () => {
    const { note, updateNoteThunk } = this.props
    const { title, content, category } = this.state
    const date = new Date()
    const updatedNote = {
      title,
      timestamp: date.toISOString(),
      category,
      content,
    }
    updateNoteThunk({ id: note.id, note: updatedNote })
    this.setState({
      editCategory: false,
      editTitle: false,
    })
  }

  render() {
    const { note, showModalAndSetDeleteNoteIdThunk } = this.props
    const {
      editTitle,
      editCategory,
      open,
      title,
      content,
      category,
    } = this.state

    return (
      <Article>
        <NoteHeader
          setEditTitle={this.setEditTitle}
          editTitle={editTitle}
          onBlur={this.onBlur}
          onChange={this.onChange}
          timestamp={note.timestamp}
          category={category}
          descriptions={content.substr(0, 10)}
          toggle={this.toggle}
          open={open}
          title={title}
        />
        {open && (
          <Main>
            <Timestamp>{moment(note.timestamp).fromNow()}</Timestamp>
            <Category>
              {editCategory ? (
                <EditText
                  value={category}
                  name="category"
                  onChange={this.onChange}
                  onBlur={this.onBlur}
                />
              ) : (
                <span>
                  {category}
                  <EditButton onClick={this.setEditCategory} />
                </span>
              )}
            </Category>
            <Content
              name="content"
              value={content}
              onChange={this.onChange}
              onBlur={this.onBlur}
            />
            <DeleteNoteButton
              onClick={() =>
                showModalAndSetDeleteNoteIdThunk({ deleteNoteId: note.id })
              }
            />
          </Main>
        )}
      </Article>
    )
  }
}

NoteItem.propTypes = {
  note: PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    timestamp: PropTypes.string.isRequired,
    category: PropTypes.string.isRequired,
    content: PropTypes.string.isRequired,
  }).isRequired,
  updateNoteThunk: PropTypes.func.isRequired,
  showModalAndSetDeleteNoteIdThunk: PropTypes.func.isRequired,
}

const mapDispatchToProps = {
  showModalAndSetDeleteNoteIdThunk,
  updateNoteThunk,
}

export default connect(
  null,
  mapDispatchToProps
)(NoteItem)
