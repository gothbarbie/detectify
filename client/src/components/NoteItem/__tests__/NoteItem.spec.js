import React from 'react'
import { shallow } from 'enzyme'

import Root from '../../../Root'
import NoteItem from '../NoteItem'

describe('Components/NoteItem', () => {
  it('renders', () => {
    const props = {
      note: {
        id: 'n01',
        title: 'Note Title',
        timestamp: '2018-11-18T14:41:23.630Z',
        category: 'Note Category',
        content: 'Lots of text that fills up the content area.',
      },
      updateNoteThunk: jest.fn(),
      showModalAndSetDeleteNoteIdThunk: jest.fn(),
    }
    const wrapper = shallow(
      <Root>
        <NoteItem {...props} />
      </Root>
    ).dive()

    expect(wrapper).toMatchSnapshot()
  })
})
