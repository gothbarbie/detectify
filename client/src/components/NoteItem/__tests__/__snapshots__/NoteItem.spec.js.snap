// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/NoteItem renders 1`] = `
<Connect(NoteItem)
  note={
    Object {
      "category": "Note Category",
      "content": "Lots of text that fills up the content area.",
      "id": "n01",
      "timestamp": "2018-11-18T14:41:23.630Z",
      "title": "Note Title",
    }
  }
  showModalAndSetDeleteNoteIdThunk={[MockFunction]}
  updateNoteThunk={[MockFunction]}
/>
`;
